src/llm/model.py

```diff
@@ -0,0 +1,285 @@
+
+        outputs = torch.concat(outputs, dim=-1)
+        outputs = self.W_out(outputs)
+        return outputs
+
+
+class EncoderLayer(nn.Module):
+    def __init__(
+        self,
+        model_dim: int,
+        num_heads: int,
+        expansion_dim: int,
+    ):
+        super().__init__()
+        self.multi_head_attention = MultiHeadAttention(
+            model_dim=model_dim, num_heads=num_heads
+        )
+        self.feed_forward = FeedForward(
+            model_dim=model_dim, expansion_dim=expansion_dim
+        )
+        self.layer_norm1 = nn.LayerNorm(normalized_shape=model_dim)
+        self.layer_norm2 = nn.LayerNorm(normalized_shape=model_dim)
+
+    def forward(
+        self, inputs: Float32[Tensor, "batch enc_seq_len model_dim"]
+    ) -> Float32[Tensor, "batch enc_seq_len model_dim"]:
+        residual = inputs
+
+        x = self.multi_head_attention(query=inputs, key=inputs, value=inputs)
+        x = self.layer_norm1(x) + residual
+
+        residual = x
+
+        x = self.feed_forward(x)
+        x = self.layer_norm2(x) + residual
+        return x
+
+
+class Encoder(nn.Module):
+    def __init__(
+        self,
+        model_dim: int,
+        num_layers: int,
+        num_heads: int,
+        expansion_dim: int,
+    ):
+        super().__init__()
+        self.encoder_layers = list()
+        for _ in range(num_layers):
+            self.encoder_layers.append(
+                EncoderLayer(
+                    model_dim=model_dim,
+                    num_heads=num_heads,
+                    expansion_dim=expansion_dim,
+                )
+            )
+
+    def forward(
+        self,
+        encoder_inputs: Float32[Tensor, "batch enc_seq_len model_dim"],
+    ) -> Float32[Tensor, "batch enc_seq_len model_dim"]:
+        encoder_outputs = list()
+        for encoder_layer in self.encoder_layers:
+            encoder_inputs = encoder_layer(encoder_inputs)
+            encoder_outputs.append(encoder_inputs)
+        return encoder_outputs
+
+
+class DecoderLayer(nn.Module):
+    def __init__(
+        self,
+        model_dim: int,
+        num_heads: int,
+        expansion_dim: int,
+    ):
+        super().__init__()
+        self.masked_multi_head_attention = MultiHeadAttention(
+            model_dim=model_dim, num_heads=num_heads
+        )
+        self.multi_head_attention = MultiHeadAttention(
+            model_dim=model_dim, num_heads=num_heads
+        )
+        self.feed_forward = FeedForward(
+            model_dim=model_dim, expansion_dim=expansion_dim
+        )
+        self.layer_norm1 = nn.LayerNorm(normalized_shape=model_dim)
+        self.layer_norm2 = nn.LayerNorm(normalized_shape=model_dim)
+        self.layer_norm3 = nn.LayerNorm(normalized_shape=model_dim)
+
+    def forward(
+        self,
+        decoder_inputs: Float32[Tensor, "batch dec_seq_len model_dim"],
+        encoder_outputs: Float32[Tensor, "batch enc_seq_len model_dim"],
```

Comment:
`encoder_output` here because this block will the output from just one encoder block.